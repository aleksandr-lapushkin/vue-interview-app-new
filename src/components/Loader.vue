<template>
  <div>
    <LoadingSpinner />
    <h6>Loading</h6>
  </div>
</template>

<script>
import LoadingSpinner from "./LoadingSpinner";

export default {
  name: "Loader",
  components: {
    LoadingSpinner
  }
};
</script>

<style scoped>
div {
  display: flex;
  flex-direction: column;
  justify-content: center;
  align-items: center;
}
</style>
