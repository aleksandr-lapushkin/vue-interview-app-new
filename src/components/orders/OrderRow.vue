<template>
  <tr>
    <td class="order-id">{{ order.id }}</td>
    <td class="order-title">{{ order.title }}</td>
    <td class="order-status">{{ order.status }}</td>
  </tr>
</template>

<script>
export default {
  name: "OrderRow",
  props: ["order"]
};
</script>

<style scoped>
td {
  padding: 1em;
}
tr:hover {
  background: #eee;
  transition-duration: 300ms;
  cursor: pointer;
}
</style>
