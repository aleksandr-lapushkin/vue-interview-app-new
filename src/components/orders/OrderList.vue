<template>
  <div>
    <div class="row">
      <div class="columns twelve">
        <router-link to="/new">
          <button class="button-primary">New</button>
        </router-link>
        <button v-on:click="onRefresh">Refresh</button>
      </div>
    </div>
    <div class="row">
      <div class="columns twelve">
        <table class="u-full-width" v-if="loaded">
          <thead>
            <tr>
              <th>ID</th>
              <th>Title</th>
              <th>Status</th>
            </tr>
          </thead>
          <tbody>
            <OrderRow
              v-for="order in orders"
              v-bind:key="order.id"
              v-bind:order="order"
            />
          </tbody>
        </table>
        <Loader class="columns twelve" v-if="!loaded" />
      </div>
    </div>
  </div>
</template>

<script>
import OrderRow from "./OrderRow";
import Loader from "../Loader";

export default {
  name: "OrderList",
  components: {
    OrderRow,
    Loader
  },
  data: () => {
    return {
      loaded: false,
      orders: []
    };
  },
  created: function() {
    this.fetchOrders().then(orders => {
      this.loaded = true;
      this.orders = orders;
    });
  },
  methods: {
    fetchOrders: function() {
      return this.$ordersDao.fetchAll();
    },
    onRefresh: function() {
      this.loaded = false;
      this.fetchOrders().then(orders => {
        this.loaded = true;
        this.orders = orders;
      });
    }
  }
};
</script>

<style scoped>
table {
  border-collapse: collapse;
}
</style>
