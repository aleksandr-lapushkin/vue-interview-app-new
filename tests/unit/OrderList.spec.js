import { shallowMount, RouterLinkStub } from "@vue/test-utils";
import OrderList from "../../src/components/orders/OrderList.vue";
import OrderRow from "../../src/components/orders/OrderRow.vue";
import Loader from "../../src/components/Loader.vue";
import Vue from "vue";

test("OrderList is loading", () => {
  const wrapper = shallowMount(OrderList, {
    stubs: {
      loader: Loader,
      "router-link": RouterLinkStub
    },
    mocks: {
      $ordersDao: {
        fetchAll: () => new Promise(() => {})
      }
    }
  });
  expect(wrapper.find(Loader).is(Loader)).toEqual(true);
  expect(wrapper.find(Loader).text()).toEqual("Loading");
  expect(wrapper.html()).toMatchSnapshot();
});

test("OrderList has 1 row", async () => {

  const promise = Promise.resolve([
    { id: 123, title: "hello world", status: "PROCESSING" }
  ]);
  const wrapper = shallowMount(OrderList, {
    stubs: {
      loader: Loader,
      "order-row": OrderRow,
      "router-link": RouterLinkStub
    },
    mocks: {
      $ordersDao: {
        fetchAll: () => promise
      }
    }
  });

  await Vue.nextTick();

  expect(wrapper.findAll(Loader).length).toEqual(0);
  expect(wrapper.findAll(OrderRow).length).toEqual(1);
  expect(wrapper.find(OrderRow).props("order")).toEqual({
    id: 123,
    title: "hello world",
    status: "PROCESSING"
  });
  expect(wrapper.html()).toMatchSnapshot();
});
