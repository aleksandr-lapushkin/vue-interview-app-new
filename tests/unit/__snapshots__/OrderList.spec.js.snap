// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`OrderList has 1 row 1`] = `
<div>
  <div class="row">
    <div class="columns twelve"><a><button class="button-primary">New</button></a> <button>Refresh</button></div>
  </div>
  <div class="row">
    <div class="columns twelve">
      <table class="u-full-width">
        <thead>
          <tr>
            <th>ID</th>
            <th>Title</th>
            <th>Status</th>
          </tr>
        </thead>
        <tbody>
          <orderrow-stub order="[object Object]"></orderrow-stub>
        </tbody>
      </table>
      <!---->
    </div>
  </div>
</div>
`;

exports[`OrderList is loading 1`] = `
<div>
  <div class="row">
    <div class="columns twelve"><a><button class="button-primary">New</button></a> <button>Refresh</button></div>
  </div>
  <div class="row">
    <div class="columns twelve">
      <!---->
      <div class="columns twelve">
        <loadingspinner-stub></loadingspinner-stub>
        <h6>Loading</h6>
      </div>
    </div>
  </div>
</div>
`;
